### PR TITLE
[functorch] test: try using reference_inputs in vmap tests

### PR DESCRIPTION
**Pull request by kshitij12345**
Ref https://github.com/pytorch/functorch/issues/1090

Timings:

`test_vmap_exhaustive`

After PR
```
== 1168 passed, 55 skipped, 2353 deselected, 153 xfailed in 195.07s (0:03:15) ==
```

Before PR
```
== 1134 passed, 55 skipped, 2316 deselected, 150 xfailed in 77.18s (0:01:17) ==
```

`test_op_has_batch_rule`

After PR
```
== 988 passed, 57 skipped, 2353 deselected, 331 xfailed in 144.70s (0:02:24) ==
```

Before PR
```
== 969 passed, 57 skipped, 2316 deselected, 313 xfailed in 65.86s (0:01:05) ==
```

### CODE REVIEW

File: test/functorch/test_vmap.py [hunk 1]
```diff
@@ -3361,11 +3361,45 @@ def test():
                             vmap(op, in_dims)(*args, **kwargs)
 
             # Sample inputs check
-            sample_inputs_itr = op.sample_inputs(device, dtype, requires_grad=False)
+            sample_inputs_op = {
+                                # Take too long
+                                "special.chebyshev_polynomial_t",
```
Comment (kshitij12345) [inline test/functorch/test_vmap.py:3366]: These ops already have skip for taking long time with reference inputs

Eg.
https://github.com/pytorch/pytorch/blob/39d49dbe4546c9f4d153ec748f62371e4546158a/torch/testing/_internal/opinfo/definitions/special.py#L374-L388

**Comment by kshitij12345** [comment]
Will take care of ASAN failure post the review.

File: test/functorch/test_vmap.py [hunk 2]
```diff
@@ -3473,11 +3496,34 @@ def test():
...
+        xfail('movedim'),
+        # RuntimeError: NYI: querying is_contiguous inside of vmap for
+        # memory_format other than torch.contiguous_format
+        xfail('contiguous'),
+        # RuntimeError: NYI: Tensor.clone(memory_format) inside vmap is only supported
+        # with memory_format torch.preserve_format or torch.contiguous_format (got ChannelsLast)
+        xfail('clone'),
+        # RuntimeError: When vmap-ing torch.nn.functional.one_hot,
+        # please provide an explicit positive num_classes argument.
+        xfail('nn.functional.one_hot'),
```
Comment (zou3519) [inline test/functorch/test_vmap.py:3513]: Normally I'd feel bad about adding these xfails, but we do have manual tests for contiguous, clone, one_hot, sub, in the codebase; and movedim is tested just by virtue of being a part of the vmap implementation.

File: test/functorch/test_vmap.py [hunk 3]
```diff
@@ -3661,6 +3711,25 @@ def test_vmap_exhaustive(self, device, dtype, op):
...
+        xfail('ge'),
+        xfail('gt'),
+        xfail('le'),
+        xfail('lt'),
+        xfail('ne'),
+        # AssertionError
+        # Mismatched elements: 18 / 20 (90.0%)
+        # Greatest absolute difference: 14.031710147857666 at index (0, 5) (up to 0.0001 allowed)
+        # Greatest relative difference: 2.9177700113052603 at index (0, 3) (up to 0.0001 allowed)
+        xfail('narrow_copy', device_type='cpu'),
```
Comment (zou3519) [inline test/functorch/test_vmap.py:3732]: Can you file an issue for silent correctness? Also, do you know which of the following is the actual problem?
- the non-contiguous test is failing
- the batching rule is bogus?
- narrow_copy has inconsistent semantics on cpu/cuda?

File: test/functorch/test_vmap.py [hunk 4]
```diff
@@ -3361,11 +3361,45 @@ def test():
...
+                "fft.fftn",
+                "fft.hfft",
+                "fft.hfft2",
+                "fft.hfftn",
+                "fft.ifft",
+                "fft.ifft2",
+                "fft.ifftn",
+                "fft.ihfft",
+                "fft.ihfft2",
+                "fft.ihfftn",
```
Comment (zou3519) [inline test/functorch/test_vmap.py:3390]: This sounds concerning, is it? (if so, file an issue please)

**Review by zou3519 — APPROVED** [review]
LGTM. We should try to dig into if some of the failures are important and file issues for them if so

File: test/functorch/test_vmap.py (hunk 3 above)
Comment (kshitij12345) [reply test/functorch/test_vmap.py:3721]: Sure, will file an issue.

* I don't think `non-contiguous` sample is an issue as we haven't added `non-contig` testing to vmap tests.
* Batching rule for `narrow_copy` seems innocuous and doesn't have special handling for CPU and CUDA.

So maybe the operator has some issue.

Batching Rule Ref:
https://github.com/pytorch/pytorch/blob/3120054c151d37fdf43963dbb60ab420908f48cf/aten/src/ATen/functorch/BatchRulesViews.cpp#L506-L515

Comment (zou3519) [reply test/functorch/test_vmap.py:3724]: If the operator is a problem: if we can come up with some repro that doesn't involve vmap that shows that on the same input (on cpu/cuda with the same strides), it produces different outputs, then that would be great. One idea to "get rid of the vmap" is to use make_fx to trace out what's happening

Comment (kshitij12345) [reply test/functorch/test_vmap.py:3724]: Sure. Thanks! 

Have assigned the issue to myself. Will have a look soon.

**Pull request closed without merging**